### ISSUE TITLE
rewrite can work on dependent goals, but this feature is not documented

### ISSUE THREAD

**Issue by anton-trunov** [coq/coq#10470]
#### Description of the problem

`rewrite` tactic supports dependent rewrites, as in the following snippet

```coq
Lemma dep_rewrite (T : Type) (x a : T) (x_is_a : x = a) :
  exist (fun x => eq x a) x x_is_a = exist _ a eq_refl.
Proof. now rewrite x_is_a. Qed.
```

but it appears this ability is not currently documented. @herbelin mentions [here](https://github.com/coq/coq/issues/5644) that in such dependent setting
> a dependent rewriting lemma is used

Related issue: #10468 (CC: @gares)

#### Coq Version

Coq 8.10beta

**Comment by herbelin** [comment]
@anton-trunov: You're right. Then, don't hesitate to propose a paragraph in the documentation. It does not have to be long.

**Comment by gares** [comment]
I guess someone knowing the code should help.

 Eg when does this feature trigger exactly? Which indexes are picked when this occurs (is `at` about the LHS or about the proof?). Also when should be this used in place of destruct?

**Comment by herbelin** [comment]
> Eg when does this feature trigger exactly? 

When the argument of `rewrite` occurs in the statement to rewrite (see `Equality.leibniz_rewrite_ebindings_clause ` for the test of dependency and `Equality.find_elim` for the selection of the rewriting scheme).

> Which indexes are picked when this occurs (is at about the LHS or about the proof?).

The occurrences of the LHS and of the proofs are picked by the unification algorithm (`Unification.w_unify`).

> Also when should be this used in place of destruct?

Using `destruct` for rewriting is a bit of an accidental side effect of `destruct`. For instance, `destruct` rewrites only from right-to-left and there is no general support for `destruct` in an hypothesis. The "dependent" feature of `rewrite` works both from right-to-left and left-to-right as well as in conclusion and in hypotheses. (Which does not mean that it cannot be improved further though, improving the selection of occurrences for instance.) So, I would not recommend users to use `destruct` for rewriting as this is not the right "level of abstraction".
